go/mobile/async.go

```diff
@@ -0,0 +1,214 @@
+package iden3mobile
+
+import (
+	"encoding/json"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/iden3/go-iden3-core/db"
+	log "github.com/sirupsen/logrus"
+)
+
+type (
+	Event interface {
+		OnEvent(string, string, string, error) // Type, Ticket ID, Data(json), error
+	}
+
+	Callback interface {
+		VerifierResponse(bool, error)
+		RequestClaimResponse(*Ticket, error)
+	}
+
+	ticketInterface interface {
+		isDone(*Identity) (bool, string, error)
+		marshal() ([]byte, error)
+		unmarshal([]byte) error
+	}
+
+	Ticket struct {
+		Id                   string
+		LastChecked          int64
+		Type                 string
+		handler              ticketInterface
+		HandlerMarshaledData []byte
+	}
+)
+
+func (i *Identity) addTicket(t *Ticket) {
+	i.Tickets.Lock()
+	i.Tickets.m[t.Id] = t
+	i.Tickets.Unlock()
+	log.Info("Ticket added: Type: ", t.Type, ". ID: ", t.Id)
+}
+
+func (i *Identity) checkPendingTickets(checkPendingTicketsPeriod time.Duration) {
+	// TODO: give more control to native host (check now, ...): Impl ctx context, Check out futures @ rust for inspiration.
+	for {
+		var wg sync.WaitGroup
+		i.Tickets.Lock()
+		// Stop the loop?
+		if i.Tickets.shouldStop {
+			log.Info("Stopping check pending tickets routine")
+			i.Tickets.Unlock()
+			return
+		}
+		// Check tickets
+		log.Info("Checking ", len(i.Tickets.m), " pending tickets")
+		finishedTickets := make([]string, len(i.Tickets.m))
+		index := 0
+		for _, t := range i.Tickets.m {
+			// Check ticket
+			wg.Add(1)
+			go func(t *Ticket, index int) {
+				defer wg.Done()
+				isDone, data, err := t.handler.isDone(i)
+				if isDone {
+					// Resolve ticket
+					log.Info("Sending event for ticket: " + t.Id)
+					finishedTickets[index] = t.Id
+					i.eventSender.OnEvent(t.Type, t.Id, data, err)
+				} else {
+					if err != nil {
```

Comment:
A ticket may return an error that means it will never succeed.  With the current implementation the ticket will be checked forever.  Maybe in the future it would be good to have three outcomes of polling (`isDone`) a ticket:
- Resolved<Data>
- Pending
- Error

Reply:
The way it's implemented, isDone returns `bool, string, error`. The bool indicates if the ticket should be removed. That way handlers have the control and can implement things like "remove me after 5 consecutive fails".